**Distance data cleaning - July 28, 2026**

In [85]:
%pip install pandas matplotlib numpy scikit-learn psycopg2-binary

Note: you may need to restart the kernel to use updated packages.


In [86]:
%pip install python-dotenv

Note: you may need to restart the kernel to use updated packages.


In [87]:
#importing the database
from src.database_service import DatabaseService


In [88]:
db = DatabaseService() 

In [89]:
dir(db)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__firstlineno__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__static_attributes__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'connect',
 'connection',
 'dbname',
 'disconnect',
 'extract_from_database',
 'get_audio_records',
 'host',
 'output_dir',
 'password',
 'port',
 'remove_test_users',
 'user']

In [90]:
db.connect() #connecting to the DB

True

app_user_id 1, 2, 3, 10, 43 are test cases and are automatically removed from the DB.

In [91]:
tables = db.extract_from_database("information_schema.tables")
tables

/Users/pegah/Projects/lemurs-mqp/src/database_service.py:194: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, self.connection)
/Users/pegah/Projects/lemurs-mqp/src/database_service.py:218: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  test_users_df = pd.read_sql(query, self.connection)
2026-07-28 07:41:47,436 - INFO - Identified 5 test users to remove


,table_catalog,table_schema,table_name,table_type,self_referencing_column_name,reference_generation,user_defined_type_catalog,user_defined_type_schema,user_defined_type_name,is_insertable_into,is_typed,commit_action
0,lemurs,public,app_user,BASE TABLE,None,None,None,None,None,YES,NO,None
1,lemurs,public,umass_id,BASE TABLE,None,None,None,None,None,YES,NO,None
2,lemurs,public,user_info,BASE TABLE,None,None,None,None,None,YES,NO,None
3,lemurs,public,auth_microsoft,BASE TABLE,None,None,None,None,None,YES,NO,None
4,lemurs,public,authorized_email,BASE TABLE,None,None,None,None,None,YES,NO,None
...,...,...,...,...,...,...,...,...,...,...,...,...
243,lemurs,public,danger_alert,BASE TABLE,None,None,None,None,None,YES,NO,None
244,lemurs,public,all_danger_alert,VIEW,None,None,None,None,None,NO,NO,None
245,lemurs,public,missing_daily_surveys,VIEW,None,None,None,None,None,NO,NO,None
246,lemurs,public,danger_alert_trigger,BASE TABLE,None,None,None,None,None,YES,NO,None


In [92]:
#looking at the table names
tables["table_name"].tolist()

['app_user',
 'umass_id',
 'user_info',
 'auth_microsoft',
 'authorized_email',
 'authorized_email_elevated',
 'app_role',
 'data',
 'pg_statistic',
 'pg_type',
 'danger_alert_email',
 'question',
 'survey',
 'survey_question',
 'survey_question_view',
 'pg_foreign_table',
 'survey_response',
 'pg_authid',
 'pg_shadow',
 'answer',
 'progress',
 'pg_roles',
 'incentive',
 'goal_progress',
 'goal',
 'pg_statistic_ext_data',
 'survey_availability',
 'calorie',
 'pg_hba_file_rules',
 'pg_settings',
 'pg_file_settings',
 'pg_backend_memory_contexts',
 'pg_ident_file_mappings',
 'speed',
 'pg_config',
 'pg_shmem_allocations',
 'pg_tables',
 'screentime',
 'screentime_app',
 'audio',
 'audio_response',
 'combined_data',
 'pg_user_mapping',
 'pg_statio_all_sequences',
 'pg_replication_origin_status',
 'pg_subscription',
 'pg_attribute',
 'pg_proc',
 'pg_class',
 'pg_attrdef',
 'pg_statio_sys_sequences',
 'pg_statio_user_sequences',
 'pg_constraint',
 'pg_inherits',
 'pg_index',
 'pg_operator',

Distance table setup

In [93]:
distance = db.extract_from_database("distance")
distance.head()
distance.shape

/Users/pegah/Projects/lemurs-mqp/src/database_service.py:194: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, self.connection)
/Users/pegah/Projects/lemurs-mqp/src/database_service.py:218: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  test_users_df = pd.read_sql(query, self.connection)
2026-07-28 07:41:47,676 - INFO - Identified 5 test users to remove
2026-07-28 07:41:47,685 - INFO - Removed 1770 records associated with test users


(10501, 10)

Shape: (10501, 10)


In [94]:
distance.columns

Index(['id', 'app_user_id', 'distance', 'start_timestamp', 'end_timestamp',
       'app_source', 'recorded_date', 'type', 'additional_data', 'status'],
      dtype='object')

In [95]:
distance.head(10)

,id,app_user_id,distance,start_timestamp,end_timestamp,app_source,recorded_date,type,additional_data,status
0,1,6,31,2025-11-18 18:03:36.149,2025-11-18 18:05:22.150,androidx.health.connect.client.records.metadat...,2025-11-18 18:09:35.230,distance,{},0
1,2,7,27,2025-11-24 19:37:10.371,2025-11-24 19:38:52.372,androidx.health.connect.client.records.metadat...,2025-11-24 19:48:51.635,distance,{},0
2,3,7,13,2025-11-24 19:38:52.371,2025-11-24 19:39:09.138,androidx.health.connect.client.records.metadat...,2025-11-24 19:48:52.144,distance,{},0
3,4,7,18,2025-11-24 19:39:09.137,2025-11-24 19:39:50.102,androidx.health.connect.client.records.metadat...,2025-11-24 19:48:52.528,distance,{},0
4,5,7,35,2025-11-24 19:39:50.101,2025-11-24 19:40:50.675,androidx.health.connect.client.records.metadat...,2025-11-24 19:48:52.922,distance,{},0
5,6,7,0,2025-11-24 19:40:50.674,2025-11-24 19:40:53.041,androidx.health.connect.client.records.metadat...,2025-11-24 19:48:53.295,distance,{},0
6,7,7,7,2025-11-26 14:42:36.385,2025-11-26 14:42:43.885,androidx.health.connect.client.records.metadat...,2025-11-26 14:53:53.475,distance,{},0
7,8,19,70,2025-12-03 13:03:06.504,2025-12-03 13:04:01.537,androidx.health.connect.client.records.metadat...,2025-12-03 13:55:38.244,distance,{},0
8,9,19,22,2025-12-03 12:51:54.220,2025-12-03 12:53:06.221,androidx.health.connect.client.records.metadat...,2025-12-03 13:55:38.603,distance,{},0
9,10,19,44,2025-12-03 13:04:01.536,2025-12-03 13:04:35.732,androidx.health.connect.client.records.metadat...,2025-12-03 13:55:38.938,distance,{},0


distance table data cleaning

1)

Data Cleaning to delete the rows with same start and end timestamps if any exists

In [96]:
same_time = distance[
    distance["start_timestamp"] == distance["end_timestamp"]
]

print(f"Number of records with same start and end timestamps: {len(same_time)}")



Number of records with same start and end timestamps: 0


In [97]:
distance = distance[
    distance["start_timestamp"] != distance["end_timestamp"]
].copy()

In [98]:
distance.shape

(10501, 10)

In [99]:
import pandas as pd
#converting start_timestamp and end_timestamp to datetime
distance["start_timestamp"] = pd.to_datetime(distance["start_timestamp"])
distance["end_timestamp"] = pd.to_datetime(distance["end_timestamp"])
distance[["start_timestamp", "end_timestamp"]].head()
#distance.dtypes

,start_timestamp,end_timestamp
0,2025-11-18 18:03:36.149,2025-11-18 18:05:22.150
1,2025-11-24 19:37:10.371,2025-11-24 19:38:52.372
2,2025-11-24 19:38:52.371,2025-11-24 19:39:09.138
3,2025-11-24 19:39:09.137,2025-11-24 19:39:50.102
4,2025-11-24 19:39:50.101,2025-11-24 19:40:50.675


In [100]:
distance.shape

(10501, 10)

In [101]:
#calculating duration in minutes (start-end timestamps)

distance['duration'] = (distance['end_timestamp'] - distance['start_timestamp']).dt.total_seconds() / 60
distance[['start_timestamp', 'end_timestamp', 'duration']].head()


,start_timestamp,end_timestamp,duration
0,2025-11-18 18:03:36.149,2025-11-18 18:05:22.150,1.766683
1,2025-11-24 19:37:10.371,2025-11-24 19:38:52.372,1.700017
2,2025-11-24 19:38:52.371,2025-11-24 19:39:09.138,0.279450
3,2025-11-24 19:39:09.137,2025-11-24 19:39:50.102,0.682750
4,2025-11-24 19:39:50.101,2025-11-24 19:40:50.675,1.009567


---------------------------------------------------------

2.

Remove app_user_id = 21, 22, 44 (globally dropped users)

In [102]:
print((distance["app_user_id"] == 44).sum())

441


In [103]:
distance.shape

(10501, 11)

In [104]:
distance = distance[distance["app_user_id"] != 44].copy()

In [105]:
distance.shape

(10060, 11)

In [106]:
print((distance["app_user_id"] == 21).sum())

1


In [107]:
distance = distance[distance["app_user_id"] != 21].copy()

In [108]:
print((distance["app_user_id"] == 22).sum())

440


In [109]:
distance = distance[distance["app_user_id"] != 22].copy()

In [110]:
distance.shape

(9619, 11)

--------------------------------------------------------

3)

Identify cases with start_time stamp before september 2025 (study timeframe)


In [111]:
pre2025=distance[distance["start_timestamp"] < "2025-09-01"][["start_timestamp", "end_timestamp", "duration", "distance", "app_user_id", "app_source"]]




In [112]:
distance[distance["start_timestamp"] < "2025-01-01"][["start_timestamp", "end_timestamp", "duration", "distance", "app_user_id", "app_source"]]

,start_timestamp,end_timestamp,duration,distance,app_user_id,app_source


In [113]:
pre2025.groupby("app_user_id").size().sort_values(ascending=False)

Series([], dtype: int64)

In [114]:
#remove cases with start_time stamp before september 2025
distance_cleaned = distance[distance["start_timestamp"] >= "2025-09-01"]
distance_cleaned.shape

(9619, 11)

--------------------------------------------------------------

4)

Removing duplicate rows having the same [app_user_id, start_timestamp, end_timestamp, distance, app_source]


In [115]:
duplicate_count = distance_cleaned.duplicated(
    subset=[
        "app_user_id",
        "start_timestamp",
        "end_timestamp",
        "distance",
        "app_source"
    ]
).sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 657


In [116]:
duplicates = distance_cleaned[
    distance_cleaned.duplicated(
        subset=[
            "app_user_id",
            "start_timestamp",
            "end_timestamp",
            "distance",
            "app_source"
        ],
        keep=False
    )
]

duplicates.sort_values(
    ["app_user_id", "start_timestamp"]
)

,id,app_user_id,distance,start_timestamp,end_timestamp,app_source,recorded_date,type,additional_data,status,duration
1211,1212,6,132,2025-12-19 16:22:17.470,2025-12-19 16:29:35.471,androidx.health.connect.client.records.metadat...,2025-12-19 17:36:45.721,distance,{},0,7.300017
1212,1213,6,132,2025-12-19 16:22:17.470,2025-12-19 16:29:35.471,androidx.health.connect.client.records.metadat...,2025-12-19 17:37:10.474,distance,{},0,7.300017
352,353,9,55,2025-12-08 09:27:09.369,2025-12-08 09:27:51.642,androidx.health.connect.client.records.metadat...,2025-12-08 15:45:50.234,distance,{},0,0.704550
434,435,9,55,2025-12-08 09:27:09.369,2025-12-08 09:27:51.642,androidx.health.connect.client.records.metadat...,2025-12-08 16:01:02.604,distance,{},0,0.704550
487,488,9,55,2025-12-08 09:27:09.369,2025-12-08 09:27:51.642,androidx.health.connect.client.records.metadat...,2025-12-08 16:11:05.591,distance,{},0,0.704550
...,...,...,...,...,...,...,...,...,...,...,...
11019,11052,48,6,2026-05-25 07:42:13.490,2026-05-25 07:42:16.064,iPhone,2026-05-25 10:48:55.862,distance,{},0,0.042900
11022,11055,48,26,2026-05-25 08:06:18.734,2026-05-25 08:08:37.820,iPhone,2026-05-25 11:08:39.819,distance,{},0,2.318100
11023,11056,48,26,2026-05-25 08:06:18.734,2026-05-25 08:08:37.820,iPhone,2026-05-25 11:08:39.825,distance,{},0,2.318100
11980,12013,49,264,2026-07-02 11:58:03.878,2026-07-02 12:36:59.078,iPhone,2026-07-02 12:51:19.464,distance,{},0,38.920000


In [117]:
duplicates.to_csv("cleaned_duplicate_rows_distance.csv", index=False)

In [118]:
distance_cleaned = distance_cleaned.drop_duplicates(
    subset=[
        "app_user_id",
        "start_timestamp",
        "end_timestamp",
        "distance",
        "app_source"
    ],
    keep="first"
)

print("Shape after removing duplicates:", distance_cleaned.shape)

Shape after removing duplicates: (8962, 11)


In [119]:
#checking if we have any duplicates now:
duplicate_rows = distance_cleaned[
    distance_cleaned.duplicated(
        subset=[
            "app_user_id",
            "start_timestamp",
            "end_timestamp",
            "distance",
            "app_source"
        ],
        keep=False
    )
]

print("Number of duplicate rows:", len(duplicate_rows))
duplicate_rows

Number of duplicate rows: 0


,id,app_user_id,distance,start_timestamp,end_timestamp,app_source,recorded_date,type,additional_data,status,duration


In [120]:
distance_cleaned.shape

(8962, 11)

-----------------------------------------------------------------------------

The final cleaned distance data (including android, apple watch, iphone) has the shape of (8962, 11) up to the date July 8, 2026.

In [126]:
print("Records:", len(distance_cleaned))
print("Users:", distance_cleaned["app_user_id"].nunique())
distance_cleaned["app_user_id"].unique()

Records: 8962
Users: 27


array([ 6,  7, 19,  9, 20, 23, 24, 25, 26, 27, 29, 28, 30, 31, 32, 33, 35,
       36, 37, 38, 39, 40, 41, 46, 47, 48, 49])

---------------------------------------------------------------
-----------------------------------------------------------

**Creating different datasets for modeling:**

('distance_cleaned' includes all the records after data cleaning)

Creating datasets including:

1. Only Android users
2. Only iPhone records (excluding Apple Watch records for users with both iPhone and Apple Watch)
3. Only IOS users (including Apple Watch records for users with both iPhone and Apple Watch, only exclude Android)

In [122]:
#1.only android users
android_distance = distance_cleaned[
    distance_cleaned["app_source"].str.contains("android", case=False, na=False)
]

print("Records:", len(android_distance))
print("Users:", android_distance["app_user_id"].nunique())
android_distance["app_user_id"].unique()

Records: 927
Users: 5


array([ 6,  7, 19,  9, 20])

----------------------------------------------------

In [123]:
#2. Only iPhone users (excluding Apple Watch records for users with both iPhone and Apple Watch)


In [124]:
# Keep only iPhone records (for all users)
iphone_only_distance = distance_cleaned[
    distance_cleaned["app_source"] == "iPhone"
]

print("Records:", len(iphone_only_distance))
print("Users:", iphone_only_distance["app_user_id"].nunique())
iphone_only_distance["app_user_id"].unique()

Records: 6844
Users: 22


array([23, 24, 25, 26, 27, 29, 28, 30, 31, 32, 33, 35, 36, 37, 38, 39, 40,
       41, 46, 47, 48, 49])

------------------------------------------------------

In [125]:
#3. Only IOS users (including Apple Watch records for users with both iPhone and Apple Watch, only exclude Android)

ios_distance = distance_cleaned[
    distance_cleaned["app_source"].isin(["iPhone", "Apple Watch"])
]

print("Records:", len(ios_distance))
print("Users:", ios_distance["app_user_id"].nunique())
ios_distance["app_user_id"].unique()

Records: 8035
Users: 22


array([23, 24, 25, 26, 27, 29, 28, 30, 31, 32, 33, 35, 36, 37, 38, 39, 40,
       41, 46, 47, 48, 49])

--------------------------------------------------------

**All datasets created (4 total):**

1) 'distance_clean': includes all the records after data cleaning (all Android and IOS (iPhone and Apple Watch records))
    - Records: 8962 - Users: 27
    - app_user_id: [ 6,  7, 19,  9, 20, 23, 24, 25, 26, 27, 29, 28, 30, 31, 32, 33, 35, 36, 37, 38, 39, 40, 41, 46, 47, 48, 49]

2) 'android_distance': includes only Android users 
    - Records: 927 - Users: 5
    - app_user_id: [6, 7, 9, 19, 20]

3) 'iphone_only_distance': includes only iPhone records (excluding Apple Watch records for users with both iPhone and Apple Watch)
    - Records: 6844 - Users: 22
    - app_user_id: [23, 24, 25, 26, 27, 29, 28, 30, 31, 32, 33, 35, 36, 37, 38, 39, 40, 41, 46, 47, 48, 49]

4) 'ios_distance': includes only IOS users (including Apple Watch records for users with both iPhone and Apple Watch, only exclude Android)
    - Records: 8035 - Users: 22
    - app_user_id: [23, 24, 25, 26, 27, 29, 28, 30, 31, 32, 33, 35, 36, 37, 38, 39, 40, 41, 46, 47, 48, 49]



--------------------------------------
-------------------------------------